ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -97,30 +126,78 @@ static INT_PTR CALLBACK ShowInfoProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPA
   return DefWindowProc(hwndDlg, uMsg, wParam, lParam);
 }
 
-static INT_PTR CALLBACK GetStrProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+/* < Used by InputBoxProc > */
+void ClientResize(HWND hWnd, int nWidth, int nHeight)
 {
-  if (uMsg == WM_INITDIALOG) {
-    SetWindowText(hwndDlg, gs_cap.c_str());
-    SetDlgItemText(hwndDlg, 12, gs_def.c_str());
-    SetDlgItemText(hwndDlg, 13, gs_message.c_str());
-  }
+  RECT rcClient, rcWind;
+  POINT ptDiff;
+  GetClientRect(hWnd, &rcClient);
+  GetWindowRect(hWnd, &rcWind);
+  ptDiff.x = (rcWind.right - rcWind.left) - rcClient.right;
+  ptDiff.y = (rcWind.bottom - rcWind.top) - rcClient.bottom;
+  MoveWindow(hWnd, rcWind.left, rcWind.top, nWidth + ptDiff.x, nHeight + ptDiff.y, TRUE);
+}
+
+WCHAR wstrPromptStr[4096];
+WCHAR wstrTextEntry[MAX_PATH + 1];
+bool HideInput = 0;
+/* < / Used by InputBoxProc > */
+
+/* < Used by get_string, get_password, get_integer, and get_passcode > */
+LRESULT CALLBACK InputBoxProc(HWND hWndDlg, UINT Msg, WPARAM wParam, LPARAM lParam)
+{
+  switch (Msg)
+  {
+  case WM_INITDIALOG:
+    {
+      ClientResize(hWndDlg, 357, 128);
+      RECT rect;
+      GetWindowRect(hWndDlg, &rect);
+      MoveWindow(hWndDlg,
+        (GetSystemMetrics(SM_CXSCREEN) / 2) - ((rect.right - rect.left) / 2),
+        (GetSystemMetrics(SM_CYSCREEN) / 3) - ((rect.bottom - rect.top) / 3),
+        rect.right - rect.left, rect.bottom - rect.top, TRUE);
+      MoveWindow(GetDlgItem(hWndDlg, IDOK), 272, 10, 75, 23, TRUE);
+      MoveWindow(GetDlgItem(hWndDlg, IDCANCEL), 272, 39, 75, 23, TRUE);
+      MoveWindow(GetDlgItem(hWndDlg, IDC_EDIT), 11, 94, 336, 23, TRUE);
+      MoveWindow(GetDlgItem(hWndDlg, IDC_PROMPT), 11, 11, 252, 66, TRUE);
+      SetWindowLongPtr(hWndDlg, GWL_EXSTYLE, GetWindowLongPtr(hWndDlg, GWL_EXSTYLE) | WS_EX_DLGMODALFRAME);
```

Comment:
We've been over this before.

Reply:
Not just about lines, it's about the functionality, the compiled dialog would possibly load a lot faster.

Reply:
@RobertBColton "the compiled dialog would possibly load a lot faster"? It already loads instantaneously. I can't imagine it getting any faster.

Reply:
You aren't benchmarking it (you don't have to, don't bother) but for all we know your dialog code could actually be twice as slow as the RC file would be.

You wouldn't notice unless you created many of these dialogs at once... and then you'd notice the RC-based one possibly being faster. But again, I can't say without a proper benchmark. Algorithm complexity is about how algorithms scale, not how they handle just a single input/dialog construction.

Perceived lag with the dialog could also occur because of contention for other resources, like the CPU. For example, if your game is doing something really logic heavy... the context switch to construct and show your dialog could become noticeably longer than the RC-based version as well.

Finally, another reason I detest the dynamic non-RC way is because it's more verbose. It takes a lot more characters to describe the same dialog that way and thus makes it harder to read. I am being serious about this, I am actually having a hard time looking at all of the `MoveWindow` calls and understanding exactly how the dialog is organized. The RC files are arguably easier to read.

Reply:
If your argument is the code is more verbose, that I understand. What I dont understand is that I have tested most of these dialog functions and the only ones that don't open instantaneously for me are the get_open/save_filename/_ext() functions, and those functions are compiled dialogs that do not use MoveWindow at all.

Reply:
Yeah, the verbosity is my main argument, there's other irrelevant advantages to RC files, but mainly, I just don't want to look at all the `MoveWindow` repeatedly, that much is ugly.

Also, the file dialogs are more complex for one and also the main reason they are slow is because of your hard drive mostly. Listing files is not as fast as you think.

Reply:
It's funny how file dialogs open instantaneously on Linux but Windows and Mac have a noticeable waiting time

Reply:
That's because Windows is NTFS 😄 I bet if you repeated that to Josh he'd make a crack at Windows about it.